src/v3/plugins/github/graphView.js

```diff
@@ -0,0 +1,274 @@
+    // Check that each edge's src has an expected prefix
+    // Check that each edge's dst has an expected prefix
+    type EdgeInvariant = {|
+      +prefix: GE.RawAddress,
+      +srcPrefixes: NodeAddressT[],
+      +dstPrefixes: NodeAddressT[],
+    |};
+    const edgeTypeToInvariants: {[type: string]: EdgeInvariant} = {
+      [GE.HAS_PARENT_TYPE]: {
+        prefix: GE._Prefix.hasParent,
+        srcPrefixes: [
+          GN._Prefix.issue,
+          GN._Prefix.pull,
+          GN._Prefix.review,
+          GN._Prefix.comment,
+        ],
+        dstPrefixes: [
+          GN._Prefix.repo,
+          GN._Prefix.issue,
+          GN._Prefix.pull,
+          GN._Prefix.review,
+        ],
+      },
+      [GE.MERGED_AS_TYPE]: {
+        prefix: GE._Prefix.mergedAs,
+        srcPrefixes: [GN._Prefix.pull],
+        dstPrefixes: [NodeAddress.fromParts(["sourcecred", "git", "commit"])],
+      },
+      [GE.REFERENCES_TYPE]: {
+        prefix: GE._Prefix.references,
+        srcPrefixes: [
+          GN._Prefix.issue,
+          GN._Prefix.pull,
+          GN._Prefix.review,
+          GN._Prefix.comment,
+        ],
+        dstPrefixes: [
+          GN._Prefix.repo,
+          GN._Prefix.issue,
+          GN._Prefix.pull,
+          GN._Prefix.review,
+          GN._Prefix.comment,
+          GN._Prefix.userlike,
+        ],
+      },
+      [GE.AUTHORS_TYPE]: {
+        prefix: GE._Prefix.authors,
+        srcPrefixes: [GN._Prefix.userlike],
+        dstPrefixes: [
+          GN._Prefix.issue,
+          GN._Prefix.review,
+          GN._Prefix.pull,
+          GN._Prefix.comment,
+        ],
+      },
+    };
+
+    for (const edge of this._graph.edges({
+      addressPrefix: GE._Prefix.base,
+      srcPrefix: NodeAddress.empty,
+      dstPrefix: NodeAddress.empty,
+    })) {
+      const address: GE.RawAddress = (edge.address: any);
+      const structuredEdge = GE.fromRaw(address);
+      const invariants = edgeTypeToInvariants[structuredEdge.type];
+      if (invariants == null) {
+        throw new Error(
+          `Invariant: Unexpected edge type ${structuredEdge.type}`
+        );
+      }
+      const {prefix, srcPrefixes, dstPrefixes} = invariants;
+      if (!EdgeAddress.hasPrefix(edge.address, prefix)) {
+        throw new Error(
+          `Invariant: Expected edge ${stringify(
+            structuredEdge
+          )} to have prefix ${EdgeAddress.toString(prefix)}`
+        );
+      }
+      if (!srcPrefixes.some((p) => NodeAddress.hasPrefix(edge.src, p))) {
+        throw new Error(
+          `Invariant: Edge ${stringify(structuredEdge)} has bad src prefix`
+        );
+      }
+      if (!dstPrefixes.some((p) => NodeAddress.hasPrefix(edge.dst, p))) {
+        throw new Error(
+          `Invariant: Edge ${stringify(structuredEdge)} has bad dst prefix`
+        );
+      }
+    }
+
+    // Check that the special constraints of the has_parent edge are maintained:
```

Comment:
If `Hom`s is implemented as suggested above, then this becomes a simple
uniqueness check.

Reply:
Yep, rewritten.